### PR TITLE
Add implementation of Fedprox example

### PR DESCRIPTION
**Pull request by cuiboyuan**
## Description

This pull request tries to implement the FedProx framework proposed in this paper: https://proceedings.mlsys.org/paper/2020/file/38af86134b65d0f10fe33d30dd76442e-Paper.pdf

FedProx is a synchronous federated learning framework that is very similar to FedAvg except that FedProx adds a proximal term to the local objective that clients wish to minimize (i.e. the training loss function of the client). In other words, a FedProx client will train to minimize the following objective:

Capture

where  is the original loss function of clients in FedAvg,  is the weight of the client model in the current training epoch,  is the weight of the global model sent by the server to the client at the beginning of the training, and  is a penalty constant of the proximal term, a hyperparameter of FedProx.

The major problem FedProx tries to solve is the impact of non-IID local training data across clients on the global model. With the proximal term added to the loss function, the training of local models at each client would be less affected by its potentially biased dataset.

Another important feature of FedProx is that for straggling clients, the server will collect partial models from them after waiting for a certain amount of time. However, the authors did not specifically describe their approach of implementation of collecting partial models. Instead, they simulate aggregation of partial models trained by stragglers by choosing a proportion of the selected clients and making them train for fewer epochs. This simulation can be reproduced by setting `straggler_simulation: true` in the configuration file under `clients`, but when `straggler_simulation: false`, the server would just wait for all selected clients to finish, as in FedAvg.

I think this feature of FedProx can be implemented if we enable the urgent request update functionality for the synchronous mode. But in that case, we need some sort of global time clock to determine when to request updates from stragglers. I do not know whether this is a good idea and I just want to ask whether it is a desirable functionality for the Plato framework.

Reference:
Li, T., Sahu, A. K., Zaheer, M., Sanjabi, M., Talwalkar, A., & Smith, V. (2020). "Federated optimization in heterogeneous networks." Proceedings of Machine Learning and Systems, 2, 429-450.

P.S. Also fixed a bug related to the sleep simulation of clients.

## How has this been tested?

Run `python fedprox.py -c fedprox_MNIST_lenet5.yml` under directory `./examples/fedprox` without errors

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue) Fixes #
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: plato/trainers/basic.py [hunk 1]
```diff
@@ -110,8 +111,8 @@ def load_model(self, filename=None, location=None):
 
     def simulate_sleep_time(self):
         """Simulate client's speed by putting it to sleep."""
-        if not (hasattr(Config().clients, 'sleep_simulation')
-                and Config().clients.sleep_simulation):
+        if hasattr(Config().clients,
+                   'sleep_simulation') and Config().clients.sleep_simulation:
```
Comment (baochunli) [inline plato/trainers/basic.py:115]: Here, the original code was correct. `sleep_simulation` equals `true` implies that the entire training session will not actually going to sleep, and instead simulate it. In this case, we are not randomly generating the number of seconds.

Comment (cuiboyuan) [reply plato/trainers/basic.py]: Oh I see. It is now updated. Thanks!

**Comment by silviafeiwang** [comment]
Hi @cuiboyuan, 

I was recently running FedProx for performance comparison on GPU servers but met this error of tensors on different devices:  
Screen Shot 2022-04-30 at 1 25 09 AM

Changing line 20 in fedprox_trainer.py to `weights = torch.tensor([], requires_grad=False).to(Config().device())` can solve the issue.

https://github.com/TL-System/plato/blob/05b649faf4c18f176155c90b08c509870f0c763c/examples/fedprox/fedprox_trainer.py#L20 


Thanks,
Fei

**Comment by cuiboyuan** [comment]
Hi Fei,

Thanks for pointing that out! I have created a PR(#161) to fix the problem you mentioned with your suggested modification and I hope that resolves the error.

Thanks,
Boyuan
